src/main/scala/vectorpipe/vectortile/Pipeline.scala

```diff
@@ -0,0 +1,126 @@
+   *
+   * Not all data is useful in all levels of the vector tile pyramid.  When a
+   * geometry will no longer be used in any subsequent generated level, it may be
+   * desirable to drop the relevant rows from the DataFrame to limit the amount
+   * of working data.  Moreover, it may be desirable to collect multiple entities
+   * into a single, aggregated entity.  These sorts of operations are made
+   * possible by the [[reduce]] function.
+   *
+   * This function will be called once before tile generation begins at the
+   * initial zoom level.  All participating geometries will already have been
+   * keyed to the given layout, with the list of relevant keys available in the
+   * "keys" field.
+   *
+   * @param   input           A DataFrame minimally containing a "geom" field of
+   *                          JTS [[Geometry]] and a field of Array[SpatialKey]
+   *                          with the name given by the keyColumn param, below
+   * @param   layoutLevel     The layout level for the target zoom
+   * @param   keyColumn       The name of the column containing
+   *                          Array[SpatialKey] giving the list of keys that the
+   *                          geometry interacts with
+   */
+  def reduce(input: DataFrame, layoutLevel: LayoutLevel, keyColumn: String): DataFrame = input
+
+  /*
+   * Lower complexity of geometry while moving to less resolute zoom levels.
+   *
+   * While moving from finer to coarser levels of the pyramid, it may not be
+   * necessary to maintain the full level of detail of available geometries.
+   * This function is used to reduce the complexity of geometries.  The
+   * [[LayoutDefinition]] will be for the target zoom level.
+   *
+   * By default, we return the input geometry without simplification, but there
+   * is a simplifier using JTS's topology-preserving simplifier available in
+   * [[vectorpipe.vectortile.Simplify]].
+   */
+  def simplify(g: jts.Geometry, layout: LayoutDefinition): jts.Geometry = g
+
+  /**
+   * Select geometries for display at a given zoom level.
+   *
+   * VectorPipe allows geometries to be carried up the pyramid for display at a
+   * later time.  This function is used to choose the elements that will be
+   * displayed at the target zoom level.  This is useful for implementing a
+   * display schema, or for example, selection based on prominence (relying on
+   * the array of spatial keys added to the data frame in the column with the
+   * longest name fitting the regular expression '[_]+keys').
+   */
+  def select(input: DataFrame, targetZoom: Int): DataFrame = input
+
+  /**
+   * Clip geometries prior to writing to vector tiles.
+   *
+   * It may be desirable to carry only the portion of a geometry that intersects
+   * a given vector tile to keep down memory usage.  This function can be used
+   * to implement such schemes.
+   */
+  def clip(geom: jts.Geometry, key: SpatialKey, layoutLevel: LayoutLevel): jts.Geometry = geom
```

Comment:
Should there be a default `clip` function available that clips to the tile bounds (with a configurable buffer in tile "pixel" units)?

Reply:
I think I'd rather have the default behavior be the fastest, most braindead thing?  I could provide a pre-baked implementation in a companion object that does what you're suggesting and extend the comment to announce its existence.